crates/cargo-util/src/paths.rs

```diff
@@ -185,10 +185,45 @@ pub fn write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> Result<()>
 /// write_atomic uses tempfile::persist to accomplish atomic writes.
 pub fn write_atomic<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> Result<()> {
     let path = path.as_ref();
-    let mut tmp = TempFileBuilder::new()
-        .prefix(path.file_name().unwrap())
-        .tempfile_in(path.parent().unwrap())?;
+    let mut builder = TempFileBuilder::new();
+
+    builder.prefix(path.file_name().unwrap());
+
+    // On unix platforms, use the same permissions as the original file. Copy only the
+    // user/group/other read/write/execute permission bits. While we need to also set the
+    // permissions again later to bypass the umask, we still need to set the permissions here as
+    // well so that we don't create a more-permissive file than the original. The tempfile lib says
+    // "Windows and others — This setting is unsupported [...]", so we don't do anything with the
+    // permissions on non-unix platforms.
+    #[cfg(unix)]
+    let perms = {
+        use std::os::unix::fs::PermissionsExt;
+
+        if let Ok(metadata) = path.metadata() {
+            // these constants are u16 on macOS
+            let mask = u32::from(libc::S_IRWXU | libc::S_IRWXG | libc::S_IRWXO);
+            let mode = metadata.permissions().mode() & mask;
+            let perms = Permissions::from_mode(mode);
+
+            builder.permissions(perms.clone());
+            Some(perms)
+        } else {
+            None
+        }
+    };
+
+    let mut tmp = builder.tempfile_in(path.parent().unwrap())?;
     tmp.write_all(contents.as_ref())?;
+
+    // On unix platforms, we set the permissions on the newly created file. When we set the
+    // permissions above, they were subject to the umask. Now that the file is created, we can use
+    // fchmod (called by the std lib; subject to change) to set the permissions which ignores the
+    // umask so that the new file has the same permissions as the old file.
+    #[cfg(unix)]
```

Comment:
Maybe on Windows we could try [setting read-only](https://doc.rust-lang.org/stable/std/fs/struct.Permissions.html#method.set_readonly) if it was read-only?

I don't know whether it would fail or not when replacing a read-only file though.

Reply:
The standard library does not actually change any permissions on Windows. It sets the read-only attribute but from a security pov this is mostly useless as anyone with write permissions could just unset it again.

Reply:
I mean, I guess it might make sense to preserve some attributes in any case. Seems not as important as perms though.

Reply:
> I mean, I guess it might make sense to preserve some attributes in any case. Seems not as important as perms though.

True. I was thinking from that angle. Just a nice-to-have, not a blocker.

BTW, how could you always notice there is a Windows related issue happening, even when nobody pinged you?

Reply:
Ha, I do miss things. But often people mention something to me privately or on discord. In this case I just happened to be browsing new PRs and this one looked interesting so I opened it then noticed a Windows thing.

Reply:
I've added code to maintain the readonly property on non-unix platforms, but this doesn't work on Windows since unlike Linux you cannot delete or replace a read-only file on Windows. So when tempfile::persist is called it tries to replace the old read-only file, which fails with an "access denied" error.

I think there are workarounds, but we couldn't do it atomically without modifying the tempfile code or changing the readonly property on the original file. So I think it would be better to ignore Windows (and non-unix platforms). What do you think?

Reply:
That makes sense. Go ahead :)


---

BTW, when I mentioned the commit organization, I meant something like this: 
* The first commit asserts the bad behavior. CI is all green with this commit.
* The second commit fixes both bug and **test**. CI is still green, and the diff bewteen shows the test change so we're more confident it fixes the previously "bad" behavior.
* See https://github.com/rust-lang/cargo/pull/13808 as a reference.

Not really a hard requirement though.

Reply:
You should be able to delete read only files. Std has an issue open about renaming (https://github.com/rust-lang/rust/issues/123985) so that may be changed in the future. Not that it helps in the here and now.

Reply:
Ah okay. I had borrowed a Windows computer and tried to delete a read-only file (and tried replacing a read-only file) on the command line with `del` and `move` and wasn't able to. But I have no idea what Windows APIs exist or what was being used by those commands. tempfile seems to use [`MoveFileExW`](https://github.com/Stebalien/tempfile/blob/e5418bd64758d1d0444e9158005f00ca7d2bc6ee/src/file/imp/windows.rs#L106).

If you have any suggestions about how to make this work on Windows let me know. My Windows API knowledge is very old and I don't have a good way to test things on Windows other than using the CI.

Reply:
I removed the non-unix code that tries to preserve the readonly property.

> BTW, when I mentioned the commit organization, I meant something like this:

Ah sorry I misunderstood. I believe I fixed this now.